bazaarvoice/emodb
# Added configurable API key rate limiting for data store updates

## Github Issue #

None

## What Are We Doing Here?

This PR adds the ability to rate limit updates to the data store by API key.  These rate limits are enforced per-instance but are automatically replicated to all instances in the local data center.  Additionally an instance-wide rate rate limit can be applied to cap the updates from all API keys to a maximum rate per instance.

Additionally this PR also closes a hole where the API keys were being printed to the request log.  Now the API key IDs are printed instead.

## How to Test and Verify

1. Check out this PR
2. Follow the instructions in `DataStoreUpdateThrottleControlTask` to enforce rate limits.
3. Perform updates via the data store REST API and verify they conform to the defined rate limits.

## Risk

### Level 

`Low`.  Although there is a non-trivial amount of code in this PR most of it is new and does not affect the existing data store workflow.  The bulk of the changes to existing classes involve dependency injection such that if there is an error in the code the service simply won't start.

### Required Testing

`Manual`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
